jane/doc/extensions/kinds.md

```diff
@@ -0,0 +1,335 @@
+  (* ... *)
+end
+```
+
+and still pass `Array` to functors expecting a `type 'a t`, which assumes `('a :
+value)`.
+
+CR ccasinghino: I just stole the above from the layouts docs, verbatim, except
+changing the word layout to kind.  Probably we should delete it from there and
+point here?
+
+Relatedly, a `with type` constraint on a signature can fill in a type with one
+that has a more specific kind.  E.g., this is legal:
+```ocaml
+module type S_any = sig
+  type t : any
+end
+
+module type S_imm = S_any with type t = int
+```
+
+This can be particularly useful for common signatures that might be implemented
+by types with any kind (e.g., `Sexpable`).
+
+# With kinds
+
+Sometimes the kind of a type constructor depends on the kinds of the types that
+instantiate its parameters.  For example, the type `'a list` can mode cross on
+the portability axis if `'a` does.
+
+We could have a `list` whose kind is restricted to work on types that more cross
+on the portability axis, to record this fact, as in:
+```ocaml
+type ('a : value mod portable) portable_list : value mod portable
+```
+But it would be annoying to have many different list types, and we certainly
+don't want to restrict the normal `list` type to work only on a subset of
+`value`s.
+
+The solution to this problem is "with kinds": kinds that record dependencies on
+types.  The actual kind of list in our.
+```ocaml
+type 'a list : value mod non_null (contended portable many with 'a)
+```
+
+CR ccasinghino: I'm giving up on explaining with-kinds here for now. I think
+there is no way to write the kind of list in the current system? (The above is
+not legal syntax)
+
+# Non-modal kind axes
+
+## Externality
+
+The externality axis records whether all a type's values may safely be ignored
+by the GC.  This may be because they are OCaml "immediates" (values represented
+by a tagged integer), because they are unboxed types like `float#` or `int32#`,
+or because they are allocated elsewhere.
+
+The axis has two possible values, with `external < internal`.  Here, `external`
+means that all values of the type are safely ignored by the GC, and internal
+means values of the type may need to be scanned.
+
+The compiler uses the externality axis for certain runtime optimizations. In
+particular, updating a mutable reference to a type that is `external` can skip
+the write barrier (i.e., it does not need a call to `caml_modify`).
+
+In the future, we plan to make externality a mode, rather than just a property
+of types.
+
+## Nullability
+
+The nullability axis records whether `NULL` (the machine word 0) is a possible
+value of a type, and is used to support the non-allocating option `'a or_null`
+type. The axis has two possible values, with `non_null < maybe_null`.
+
+A type may be `non_null` if none of its value values are `NULL`.  Such types are
+compatible with `or_null`, whose definition is:
+```ocaml
+type ('a : value mod non_null) or_null : value mod maybe_null =
```

Comment:
`value mod maybe_null` doesn't work. Instead, I think it's `value_maybe_null`, but @dkalinichenko-js knows best.